### ISSUE TITLE
unrealized closures "collectioninstance_var.iter.map()" can be flagged

### ISSUE THREAD

**Issue by ai-mannamalai** [rust-lang/rust-clippy#16834]
### What it does

A Rust n00b issue;
Unrealized closures "collectioninstance_var.iter.map()" are compiled and the use-def chain for this map ends right at the statement; a Clippy rule can flag this if the use-def chain does not propagate to a iterator consumer or collect() operation.

### Advantage

- Catch unused iterators and most likely a logical bug for the programmer
- 

### Drawbacks

- Not sure; perhaps side-effect inspection could be problematic ?

### Example

```rust
fn do_something() -> i32 {
   let x = vec![1,2,3,4];
   x.iter().map(|i| { i*2 })
   return x[0]
}
```

Could be written as:

```rust
fn do_something() -> i32 {
   let x = vec![1,2,3,4];
   x = x.iter().map(|i| { i*2 }).collect::<i32>();
   return x[0]
}
```

### Comparison with existing lints

NA

### Additional Context

Unrealized closure iterators are logical bugs

**Comment by ai-mannamalai** [comment]
I find it similar/adjacent issue to https://github.com/rust-lang/rust-clippy/issues/6014

**Comment by ada4a** [comment]
For what it's worth, the compiler already gives a warning in this case:
```
warning: unused `Map` that must be used
 --> src/lib.rs:6:4
  |
6 |    x.iter().map(|i| { i*2 });
  |    ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: iterators are lazy and do nothing unless consumed
  = note: `#[warn(unused_must_use)]` (part of `#[warn(unused)]`) on by default
help: use `let _ = ...` to ignore the resulting value
  |
6 |    let _ = x.iter().map(|i| { i*2 });
  |    +++++++
```
I think this does pretty much exactly what you want?

**Comment by ai-mannamalai** [comment]
Okay
